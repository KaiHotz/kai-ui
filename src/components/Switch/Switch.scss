.ui-switch {
  box-sizing: border-box;
  position: relative;
  display: inline-block;
  text-align: left;
  width: calc(var(--ui-swich__slider-size) * 2);
  height: var(--ui-swich__height);
  &--with-label {
    margin-right: 8px;
  }
  &--disabled {
    opacity: 0.3;
  }
  &--small {
    --ui-swich__slider-size: 12px;
    --ui-swich__height: 8px;
  }
  &--medium {
    --ui-swich__slider-size: 20px;
    --ui-swich__height: 12px;
  }

  &__input {
    display: none;
    &:checked + .ui-switch__slider {
      background-color: var(--green-3);
      &:before {
        transform: translateX(var(--ui-swich__slider-size));
      }
    }
  }

  &__slider {
    box-sizing: border-box;
    position: absolute;
    cursor: pointer;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: var(--gray-4);
    transition: background-color 0.3s ease-out;
    border-radius: 15px;
    &:before {
      content: '';
      box-sizing: border-box;
      position: absolute;
      top: 0;
      bottom: 0;
      margin: auto 0;
      height: var(--ui-swich__slider-size);
      width: var(--ui-swich__slider-size);
      background-color: var(--gray-3);
      border: 1px solid var(--border-color);
      transition: transform 0.2s ease-out;
      border-radius: 50%;
    }
  }
}
