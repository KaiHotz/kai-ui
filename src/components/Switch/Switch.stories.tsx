import React, { ChangeEvent, useState } from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { FaCircleInfo, FaSistrix } from 'react-icons/fa6';

import { Switch } from './Switch';

const iconOptions = {
  none: undefined,
  searchIcon: <FaSistrix size={13} />,
  infoIcon: <FaCircleInfo size={13} />,
};

const iconMap = {
  none: 'None',
  searchIcon: 'Search',
  infoIcon: 'Information',
};

const meta: Meta<typeof Switch> = {
  title: 'Components/Switch',
  component: Switch,
  argTypes: {
    name: {
      control: false,
    },
    value: {
      control: false,
    },
    onChange: {
      control: false,
    },
    testId: {
      control: false,
    },
    className: {
      control: false,
    },
    labelEndAdornment: {
      options: Object.keys(iconOptions),
      mapping: iconOptions,
      control: {
        type: 'select',
        labels: iconMap,
      },
    },
  },
};

type Story = StoryObj<typeof Switch>;

export const Default: Story = {
  render: function useStory(args) {
    const [inputValue, setInputValue] = useState(false);

    const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
      const { checked } = e.target;
      setInputValue(checked);
    };

    return <Switch {...args} value={inputValue} onChange={handleChange} />;
  },
  args: {
    size: 'medium',
    label: '',
    labelPosition: 'right',
    labelEndAdornment: 'none',
    disabled: false,
  },
};

export default meta;
