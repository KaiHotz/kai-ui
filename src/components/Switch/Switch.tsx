import React, { forwardRef } from 'react';
import cx from 'clsx';

import './Switch.scss';
import { ILabelProps, Label } from '../Label';

export interface ISwitchProps {
  name?: string;
  value?: boolean;
  disabled?: boolean;
  size?: 'small' | 'medium';
  label?: string;
  labelPosition?: ILabelProps['position'];
  className?: string;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur?: (e: React.FocusEvent<HTMLInputElement>) => void;
  testId?: string;
}

export const Switch = forwardRef<HTMLInputElement, ISwitchProps>(
  (
    {
      onChange,
      disabled,
      name,
      value,
      label,
      labelPosition = 'right',
      size = 'medium',
      testId = 'ui-switch',
      className,
      ...rest
    },
    ref,
  ) => {
    return (
      <Label text={label} position={labelPosition}>
        <span
          className={cx(
            'ui-switch',
            {
              'ui-switch--disabled': disabled,
              'ui-switch--small': size === 'small',
              'ui-switch--medium': size === 'medium',
            },
            className,
          )}
          data-testid={testId}
        >
          <input
            className="ui-switch__input"
            onChange={onChange}
            ref={ref}
            {...rest}
            name={name}
            checked={value}
            disabled={disabled}
            type="checkbox"
            data-testid={`${testId}-input`}
          />
          <span className="ui-switch__slider" data-testid={`${testId}-slider`} />
        </span>
      </Label>
    );
  },
);
