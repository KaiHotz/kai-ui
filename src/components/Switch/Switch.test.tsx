import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';

import { Switch } from './Switch';

describe('<Switch />', () => {
  it('should render Switch', () => {
    render(<Switch name="name" value onChange={() => vi.fn} />);
    expect(screen.getByTestId('dm-switch')).toMatchSnapshot();
  });
});
