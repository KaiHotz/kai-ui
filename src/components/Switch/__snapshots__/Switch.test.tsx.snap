// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Switch /> > should render Switch 1`] = `
<span
  class="ui-switch ui-switch--medium"
  data-testid="ui-switch"
>
  <input
    checked=""
    class="ui-switch__input"
    data-testid="ui-switch-input"
    name="name"
    type="checkbox"
  />
  <span
    class="ui-switch__slider"
    data-testid="ui-switch-slider"
  />
</span>
`;
